src/go/rpk/pkg/cli/cmd/wasm/list.go

```diff
@@ -0,0 +1,178 @@
+	return command
+}
+
+func printResults(dataSet *map[int32]coprocNodeInfo, raw bool) {
+	if raw {
+		stringRep, err := json.Marshal(dataSet)
+		if err != nil {
+			fmt.Errorf("Internal error with list command, failed to serialize result into json: %s", err)
+		} else {
+			log.Info(string(stringRep))
+		}
+		return
+	}
+	t := ui.NewRpkTable(log.StandardLogger().Out)
+	t.SetColWidth(80)
+	t.SetAutoWrapText(false)
+	unresponsiveEngines := []int32{}
+	header := []string{"Coprocessor Name", "Node ID", "Input Topics", "Description"}
+	t.SetHeader(header[:])
+	for _, val := range *dataSet {
+		if val.Status != "up" {
+			unresponsiveEngines = append(unresponsiveEngines, val.NodeId)
+		}
+		for name, data := range val.Coprocessors {
+			topics := strings.Join(data.InputTopics, ",")
+			topics = "[" + topics + "]"
+			row := []string{name, strconv.Itoa(int(val.NodeId)), topics, data.Description}
+			t.Append(row)
+		}
+	}
+	t.Render()
+	if len(unresponsiveEngines) > 0 {
+		log.Infof("Unresponsive engines: [%v]", unresponsiveEngines)
+	}
+}
+
+func accrueResults(
+	messages []*sarama.ConsumerMessage,
+) (*map[int32]coprocNodeInfo, error) {
+	allData := make(map[int32]coprocNodeInfo)
+	for _, msg := range messages {
+		msgKey := int32(binary.LittleEndian.Uint32(msg.Key))
+		var info coprocNodeInfo
+		if err := json.Unmarshal(msg.Value, &info); err != nil {
+			return nil, err
+		}
+		if msgKey != info.NodeId {
+			return nil, fmt.Errorf("Mismatched node ids, key: %d reported %d", msgKey, info.NodeId)
+		}
+		allData[msgKey] = info
+	}
+	return &allData, nil
+}
+
+func consumeStatusUpdates(
+	pc sarama.PartitionConsumer, finalOffset int64,
+) ([]*sarama.ConsumerMessage, error) {
+	var messages []*sarama.ConsumerMessage
+	var err error
+	doneCh := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case msg := <-pc.Messages():
+				messages = append(messages, msg)
+				if msg.Offset >= (finalOffset - 1) {
+					doneCh <- struct{}{}
+				}
+			case msgerr := <-pc.Errors():
+				err = msgerr.Err
+				doneCh <- struct{}{}
+			}
+		}
+	}()
+	<-doneCh
+	return messages, err
+}
+
+func wasmList(
+	clientClosure func() (sarama.Client, error),
+	consumerClosure func(sarama.Client) (sarama.Consumer, error),
+	raw bool,
+) error {
+	client, err := clientClosure()
+	if err != nil {
+		return fmt.Errorf("Failed to make client: %s", err)
+	}
+	finalOffset, err := client.GetOffset(kafka.CoprocessorStatusTopic, 0, sarama.OffsetNewest)
+	if err != nil {
+		if raw {
+			log.Info("{\"status\": \"uninitialized\"}")
+		} else {
+			log.Error("CoprocessorStatusTopic is uninitialized")
+		}
+		return nil
+	}
+	if finalOffset == 0 {
+		if raw {
+			log.Info("{}")
+		} else {
+			log.Info("There are no coprocessor events to list")
+		}
+		return nil
+	}
+
+	consumer, err := consumerClosure(client)
+	if err != nil {
+		return fmt.Errorf("Failed to make consumer: %s", err)
+	}
+
+	pc, err := consumer.ConsumePartition(kafka.CoprocessorStatusTopic, 0, sarama.OffsetOldest)
```

Comment:
Sarama actually doesn't have an interface to read from a topic. Whats done here is similar to `rpk topic consume` from `consume.go`.

Furthermore we read from offset 0 because this topic is compacted by node_id. So reading the entire topic (and removing oldest entries with duplicate keys, which one of these go methods does) is desired.

Reply:
wicked